### PR TITLE
#1: Added Layer rules L1-L12

### LINKED ISSUE #1: Implement layer rules
Implement the [Layer-Rules](https://github.com/devonfw/sonar-devon4j-plugin/blob/master/documentation/rules.asciidoc#layer-rules) with ArchUnit.

Acceptance-Criteria:
* [x] Rules L1-L12 are implemented with ArchUnit
* [x] Create Code that violates Layer Rules

### PR DESCRIPTION
**Pull request by vlad961**
#1 
Implemented Layer rules L1-12 with Archunit

### CODE REVIEW

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java [hunk 1]
```diff
@@ -16,15 +16,14 @@ public class ArchitectureTest {
...
-          .layer("service").definedBy("..service..") //
-          .layer("batch").definedBy("..batch..") //
-          .layer("ui").definedBy("..ui..") //
-          // TODO
-          .because("Dependency of technical layers violates architecture rules.");
+          .layer("common").definedBy("com.devonfw.sample.archunit.common..") //
+          .layer("logic").definedBy("com.devonfw.sample.archunit.logic..") //
+          .layer("dataaccess").definedBy("com.devonfw.sample.archunit.dataaccess..") //
+          .layer("service").definedBy("com.devonfw.sample.archunit.service..") //
+          .layer("client").definedBy("com.devonfw.sample.archunit.client..")
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: @vlad961 the package root and especially the component should not be encoded here:
* We might have multiple components in the same application and each could have any up to all out of this layers
* As we want to provide a "starting point" and "template" for projects using devonfw and arch-unit we should not tie this to `com.devonfw.sample.archunit` as otherwise every project would have to adopt this to their own root package what is however not neccessary.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java [hunk 2]
```diff
@@ -16,15 +16,14 @@ public class ArchitectureTest {
...
 
+          .withOptionalLayers(true)
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: aha, nice that you found this one to prevent the error if layer is empty :) Great 👍

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import com.tngtech.archunit.core.importer.ImportOption;
+import com.tngtech.archunit.junit.AnalyzeClasses;
+import com.tngtech.archunit.junit.ArchTest;
+import com.tngtech.archunit.lang.ArchRule;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.classes;
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.noClasses;
+
+@AnalyzeClasses(packages = "com.devonfw.sample.archunit", importOptions = ImportOption.DoNotIncludeTests.class)
+public class LayerDependencyRulesTest {
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: If you put the rules which layer can access which other layer into this separate class the `ArchitectureTest` remains pointless as it more or less does nothing and can never fail or do I miss something?

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    
+    // 'access' catches only violations by real accesses, i.e. accessing a field, calling a method; compare 'dependOn' further down
+
+    // L09: verifying that code from logic layer does not access service layer (of same app).
+    @ArchTest
+    static final ArchRule logic_should_not_access_services =
+            noClasses().that().resideInAPackage("..logic..")
+                    .should().accessClassesThat().resideInAPackage("..service..");
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: you should follow the code sample I provided in the story with `mayOnlyBeAccessedByLayers` to utilize the concept of layer-checks in arch-unit.

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    @ArchTest
+    static final ArchRule services_should_only_be_accessed_by_clients_or_other_services =
+            classes().that().resideInAPackage("..service..")
+                    .should().onlyBeAccessed().byAnyPackage("..client..", "..service..");
+
+    @ArchTest
+    static final ArchRule services_should_only_access_logic_common_or_other_services =
+            classes().that().resideInAPackage("..service..")
+                    .should().onlyAccessClassesThat().resideInAnyPackage("..service..", "..common..", "..logic..", "java..", "javax..");
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: If we start whitelisting what packages are allowed to reference (like `java.*` and `javax.*`) we will end up in a maintenance hell. Thus the `mayOnlyBeAccessedByLayers` will make it much easier for us to only say what layer can access what other layers.

**Review by hohwille — CHANGES_REQUESTED** [review]
@vlad961 Thanks for your PR. Nice that you found the solution to prevent errors if a layer is empty. 👍 
I have added several review comments. Can you please have a look and try to address my concerns?

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java (hunk 1 above)
Comment (vlad961) [reply src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: I have changed each layer definition by exchanging the prefix: "com.devonfw.sample.archunit" with a second "." (dot).

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java (hunk 3 above)
Comment (vlad961) [reply src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: No, you did not miss anything. 
I left the ArchitectureTest in case you have planned something additional with it, except serving as a template.

With respect to your concerns, I updated ArchitectureTest and deleted LayerDependencyRulesTest since it is not needed now anymore.

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java (hunk 4 above)
Comment (vlad961) [reply src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: Thank you for the hint!

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java (hunk 5 above)
Comment (vlad961) [reply src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: In reference to your comment from above and this one:

Thank you for the hint.
This helped me to figure out why my first approach led to plenty violations of dependencies with reference to java.*, javax.*, org.mapstruct.* etc.

I've updated my code with respect to your suggestions and above violations have been resolved.
 - I deleted LayerDependencyRulesTest.java 
 - and adjusted ArchitectureTest.java to use correct layer definitions and utilizing mayOnlyBeAccessedByLayers rules.

Upon above changes I also updated my PR #14 . All 12 intended violations are detected.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java [hunk 6]
```diff
@@ -15,16 +15,23 @@ public class ArchitectureTest {
...
+      layeredArchitecture().consideringAllDependencies() //      
+      .layer("common").definedBy("..common..") //
+      .layer("logic").definedBy("..logic..") //
+      .layer("dataaccess").definedBy("..dataaccess..") //
+      .layer("service").definedBy("..service..") //
+      .layer("client").definedBy("..client..")
+      .layer("batch").definedBy("..batch..")
+
+      .whereLayer("client").mayNotBeAccessedByAnyLayer()
+      .whereLayer("batch").mayOnlyBeAccessedByLayers( "logic")
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: IMHO code dependencies from `logic` layer to `batch` are violating the layered architecture since the `batch` layer is "above" the `logic` layer just like the `service` layer:
https://github.com/devonfw/sonar-devon4j-plugin/blob/master/documentation/rules.asciidoc#architectural-dependency-rules

Typically batches are started via
* timers (like cron expressions with e.g. quartz - e.g. every 5 minutes, every night at 1:55 or every Friday at 22:00)
* external events - typically data to import was delivered

However, there are applications with architecture where batches are triggered from end-users via use-cases directly (synchronously within the same Java VM). This is not what devonfw recommends as batches should scale independent from OLTP and also batches have a tendency to cause trouble (performance drain, OutOfMemoryException) and in such case this would bring your OLTP down as well.
Further, as you can see below `batch` may depend on `logic` what introduces a cyclic dependency between layers what is discouraged. Often you simply decouple the triggering of batches from code-invocations as you simple write specific status or records into the DB that makes a batch to process that data.

To summarize, I would suggest to use `mayNotBeAccessedByAnyLayer()` as well for `batch`.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java [hunk 7]
```diff
@@ -15,16 +15,23 @@ public class ArchitectureTest {
...
+      .whereLayer("service").mayOnlyBeAccessedByLayers("client")
+      .whereLayer("logic").mayOnlyBeAccessedByLayers("service", "batch")
+      .whereLayer("dataaccess").mayOnlyBeAccessedByLayers("logic")
+      .whereLayer("common").mayOnlyBeAccessedByLayers("common", "dataaccess", "logic", "service")
```
Comment (hohwille) [inline src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: Actually `common` may be accessed by any layer. Further, always code from one layer may always be allowed to call code from the same layer on the level of layer checks (`common` should not be explicitly listed here and `batch` or `client` are missing). Only for cross-component calls there is a difference (Rules C3-C7 that are not part of this story). 

So here I would just use `mayNotAccessAnyLayer()` forbidding access from `common` to other but not vice versa.

**Review by hohwille — CHANGES_REQUESTED** [review]
@vlad961 thanks for your update. Looks almost perfect now. 👍 
Still I have two review annotations you should address.

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java (hunk 3 above)
Comment (hohwille) [reply src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: Thanks, done and resolved.

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java (hunk 4 above)
Comment (hohwille) [reply src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: Thanks, done and resolved.

File: src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java (hunk 5 above)
Comment (hohwille) [reply src/test/java/com/devonfw/sample/archunit/LayerDependencyRulesTest.java]: Thanks, done and resolved.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java (hunk 6 above)
Comment (vlad961) [reply src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: Thank you for your elaborate explanation. I appreciate it.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java (hunk 7 above)
Comment (vlad961) [reply src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: If I use mayNotAccessAnyLayer() on the `common` layer, I encounter the primary problem:
The `common` layer depends on other packages like _java.*_ and _javax.*_ and throws errors.
The mayNotAccessAnyLayer rule seems too restrictive.

~~I suggest adding batch and client layers to the mayOnlyBeAccessedByLayers() rule and excluding `common` itself~~ 
~~as mentioned by you.~~
~~With this rule no errors regarding the dependencies of `common` upon java.* and javax.* arise.~~ 

~~Alternatively, I need to ignore the given dependencies by the ignoreDependency() rule if the~~ ~~mayNotAccessAnyLayer rule is preferred.~~

Edit: 
As discussed earlier, I deleted the current `common` layer rule and postponed its implementation in case the implicit misuse given by the other layer rules isn't enough.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java (hunk 6 above)
Comment (hohwille) [reply src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: Thanks for your update and rework.

File: src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java (hunk 7 above)
Comment (hohwille) [reply src/test/java/com/devonfw/sample/archunit/ArchitectureTest.java]: Great. Lets see via tests, if that already does the job or we need to explicitly list all other layers here.

**Review by hohwille — APPROVED** [review]
@vlad961 thanks for your rework. Ready for merge. 👍
